token/token.go

```diff
@@ -1,75 +1,146 @@
+	GetKeysEndpoint() string
+	GetKeycloakDevModeURL() string
+}
+
+// TokenClaims represents access token claims
+type TokenClaims struct {
+	Name          string                `json:"name"`
+	Username      string                `json:"preferred_username"`
+	GivenName     string                `json:"given_name"`
+	FamilyName    string                `json:"family_name"`
+	Email         string                `json:"email"`
+	Company       string                `json:"company"`
+	SessionState  string                `json:"session_state"`
+	Authorization *AuthorizationPayload `json:"authorization"`
+	jwt.StandardClaims
+}
+
+// AuthorizationPayload represents an authz payload in the rpt token
+type AuthorizationPayload struct {
+	Permissions []Permissions `json:"permissions"`
+}
+
+// Permissions represents a "permissions" in the AuthorizationPayload
+type Permissions struct {
+	ResourceSetName *string `json:"resource_set_name"`
+	ResourceSetID   *string `json:"resource_set_id"`
+}
+
+type PublicKey struct {
+	KeyID string
+	Key   *rsa.PublicKey
+}
+
 // Manager generate and find auth token information
 type Manager interface {
-	Extract(string) (*account.Identity, error)
 	Locate(ctx context.Context) (uuid.UUID, error)
-	PublicKey() *rsa.PublicKey
+	ParseToken(ctx context.Context, tokenString string) (*TokenClaims, error)
+	PublicKey(kid string) *rsa.PublicKey
+	PublicKeys() []*rsa.PublicKey
 }
 
 type tokenManager struct {
-	publicKey  *rsa.PublicKey
-	privateKey *rsa.PrivateKey
+	publicKeysMap map[string]*rsa.PublicKey
+	publicKeys    []*PublicKey
 }
 
 // NewManager returns a new token Manager for handling tokens
-func NewManager(publicKey *rsa.PublicKey) Manager {
-	return &tokenManager{
-		publicKey: publicKey,
+func NewManager(config configuration) (Manager, error) {
+	// Load public keys from Auth service and add them to the manager
+	tm := &tokenManager{
+		publicKeysMap: map[string]*rsa.PublicKey{},
 	}
-}
 
-// NewManagerWithPrivateKey returns a new token Manager for handling creation of tokens with both private and pulic keys
-func NewManagerWithPrivateKey(privateKey *rsa.PrivateKey) Manager {
-	return &tokenManager{
-		publicKey:  &privateKey.PublicKey,
-		privateKey: privateKey,
-	}
-}
-
-func (mgm tokenManager) Extract(tokenString string) (*account.Identity, error) {
-	token, err := jwt.Parse(tokenString, func(token *jwt.Token) (interface{}, error) {
-		return mgm.publicKey, nil
-	})
+	remoteKeys, err := authtoken.FetchKeys(config.GetKeysEndpoint())
 	if err != nil {
-		return nil, errors.WithStack(err)
+		log.Error(nil, map[string]interface{}{
+			"keys_url": config.GetKeysEndpoint(),
+		}, "unable to load public keys from remote service")
+		return nil, errors.New("unable to load public keys from remote service")
+	}
+	for _, remoteKey := range remoteKeys {
+		tm.publicKeysMap[remoteKey.KeyID] = remoteKey.Key
+		tm.publicKeys = append(tm.publicKeys, &PublicKey{KeyID: remoteKey.KeyID, Key: remoteKey.Key})
```

Comment:
why do you need a `list` and a `map` at the same time ?

Reply:
We need the list to guarantee the order of the keys (just matter of performance to have tokens validated against the most likely used key first) and map for quick access to the keys by ID.